server/src/main/java/org/elasticsearch/search/dfs/DfsPhase.java

```diff
@@ -156,34 +158,42 @@ public CollectionStatistics collectionStatistics(String field) throws IOExceptio
 
     private void executeKnnVectorQuery(SearchContext context) throws IOException {
         SearchSourceBuilder source = context.request().source();
-        if (source == null || source.knnSearch() == null) {
+        if (source == null || source.knnSearch().isEmpty()) {
             return;
         }
 
         SearchExecutionContext searchExecutionContext = context.getSearchExecutionContext();
-        KnnSearchBuilder knnSearch = context.request().source().knnSearch();
-        KnnVectorQueryBuilder knnVectorQueryBuilder = knnSearch.toQueryBuilder();
+        List<KnnSearchBuilder> knnSearch = context.request().source().knnSearch();
+        List<KnnVectorQueryBuilder> knnVectorQueryBuilders = knnSearch.stream().map(KnnSearchBuilder::toQueryBuilder).toList();
 
         if (context.request().getAliasFilter().getQueryBuilder() != null) {
-            knnVectorQueryBuilder.addFilterQuery(context.request().getAliasFilter().getQueryBuilder());
+            for (KnnVectorQueryBuilder knnVectorQueryBuilder : knnVectorQueryBuilders) {
+                knnVectorQueryBuilder.addFilterQuery(context.request().getAliasFilter().getQueryBuilder());
+            }
         }
-
-        Query query = searchExecutionContext.toQuery(knnVectorQueryBuilder).query();
-        TopScoreDocCollector topScoreDocCollector = TopScoreDocCollector.create(knnSearch.k(), Integer.MAX_VALUE);
-        Collector collector = topScoreDocCollector;
-
+        List<DfsKnnResults> knnResults = new ArrayList<>(knnVectorQueryBuilders.size());
+        for (int i = 0; i < knnSearch.size(); i++) {
+            Query knnQuery = searchExecutionContext.toQuery(knnVectorQueryBuilders.get(i)).query();
+            TopScoreDocCollector topScoreDocCollector = TopScoreDocCollector.create(knnSearch.get(i).k(), Integer.MAX_VALUE);
+            Collector collector = topScoreDocCollector;
+            if (context.getProfilers() != null) {
+                InternalProfileCollector ipc = new InternalProfileCollector(
+                    topScoreDocCollector,
+                    CollectorResult.REASON_SEARCH_TOP_HITS,
+                    List.of()
+                );
+                QueryProfiler knnProfiler = context.getProfilers().getDfsProfiler().addQueryProfiler(ipc);
+                collector = ipc;
+                // Set the current searcher profiler to gather query profiling information for gathering top K docs
+                context.searcher().setProfiler(knnProfiler);
```

Comment:
This is a big change to how DFS gathers stats for KNN querying. We have a unique Query & Collector here per query shard when profiling. I create a new query profiler for every KNN query, set that to be the searchers profiler, and gather stats.

I then set the context's profiler to be the overall current query profiler.

```json
{
    "took": 10,
    "timed_out": false,
    "_shards": {
    },
    "hits": {
    },
    "profile": {
        "shards": [
            {
                "id": "[MD9ob8tHTrScTDnJQgX4EA][image-index][0]",
                "dfs": {
                    "statistics": {
                        "type": "statistics",
                        "description": "collect term statistics",
                        "time_in_nanos": 1248250,
                        "breakdown": {
                            "term_statistics": 3333,
                            "collection_statistics": 26708,
                            "collection_statistics_count": 2,
                            "create_weight": 1242209,
                            "term_statistics_count": 1,
                            "rewrite_count": 0,
                            "create_weight_count": 1,
                            "rewrite": 0
                        }
                    },
                    "knn": {
                        "query": [
                            {
                                "type": "BoostQuery",
                                "description": "(DocAndScore[50])^0.1",
                                "time_in_nanos": 18087,
                                "breakdown": {
                                    "set_min_competitive_score_count": 0,
                                    "match_count": 0,
                                    "shallow_advance_count": 0,
                                    "set_min_competitive_score": 0,
                                    "next_doc": 1293,
                                    "match": 0,
                                    "next_doc_count": 4,
                                    "score_count": 4,
                                    "compute_max_score_count": 0,
                                    "compute_max_score": 0,
                                    "advance": 3541,
                                    "advance_count": 3,
                                    "count_weight_count": 0,
                                    "score": 751,
                                    "build_scorer_count": 6,
                                    "create_weight": 1834,
                                    "shallow_advance": 0,
                                    "count_weight": 0,
                                    "create_weight_count": 1,
                                    "build_scorer": 10668
                                }
                            },
                            {
                                "type": "BoostQuery",
                                "description": "(DocAndScore[10])^0.5",
                                "time_in_nanos": 10791,
                                "breakdown": {
                                    "set_min_competitive_score_count": 0,
                                    "match_count": 0,
                                    "shallow_advance_count": 0,
                                    "set_min_competitive_score": 0,
                                    "next_doc": 958,
                                    "match": 0,
                                    "next_doc_count": 3,
                                    "score_count": 3,
                                    "compute_max_score_count": 0,
                                    "compute_max_score": 0,
                                    "advance": 2417,
                                    "advance_count": 3,
                                    "count_weight_count": 0,
                                    "score": 542,
                                    "build_scorer_count": 6,
                                    "create_weight": 1000,
                                    "shallow_advance": 0,
                                    "count_weight": 0,
                                    "create_weight_count": 1,
                                    "build_scorer": 5874
                                }
                            }
                        ],
                        "rewrite_time": 571958,
                        "collector": [
                            {
                                "name": "KnnQueryCollector",
                                "reason": "search_multi",
                                "time_in_nanos": 27207,
                                "children": [
                                    {
                                        "name": "SimpleTopScoreDocCollector",
                                        "reason": "search_top_hits",
                                        "time_in_nanos": 17250
                                    },
                                    {
                                        "name": "SimpleTopScoreDocCollector",
                                        "reason": "search_top_hits",
                                        "time_in_nanos": 9957
                                    }
                                ]
                            }
                        ]
                    }
                },
                "searches": [
                    {
                        "query": [
                            {
                                "type": "BooleanQuery",
                                "description": "ScoreAndDocQuery (title:mountain title:lake)^0.9",
                                "time_in_nanos": 326792,
                                "breakdown": {
                                    "set_min_competitive_score_count": 0,
                                    "match_count": 2,
                                    "shallow_advance_count": 0,
                                    "set_min_competitive_score": 0,
                                    "next_doc": 13584,
                                    "match": 417,
                                    "next_doc_count": 4,
                                    "score_count": 4,
                                    "compute_max_score_count": 0,
                                    "compute_max_score": 0,
                                    "advance": 23084,
                                    "advance_count": 3,
                                    "count_weight_count": 0,
                                    "score": 4499,
                                    "build_scorer_count": 6,
                                    "create_weight": 111292,
                                    "shallow_advance": 0,
                                    "count_weight": 0,
                                    "create_weight_count": 1,
                                    "build_scorer": 173916
                                },
                                "children": [
                                    {
                                        "type": "KnnScoreDocQuery",
                                        "description": "ScoreAndDocQuery",
                                        "time_in_nanos": 17252,
                                        "breakdown": {
                                            "set_min_competitive_score_count": 0,
                                            "match_count": 0,
                                            "shallow_advance_count": 4,
                                            "set_min_competitive_score": 0,
                                            "next_doc": 458,
                                            "match": 0,
                                            "next_doc_count": 2,
                                            "score_count": 4,
                                            "compute_max_score_count": 4,
                                            "compute_max_score": 1958,
                                            "advance": 3417,
                                            "advance_count": 5,
                                            "count_weight_count": 0,
                                            "score": 709,
                                            "build_scorer_count": 9,
                                            "create_weight": 1042,
                                            "shallow_advance": 1917,
                                            "count_weight": 0,
                                            "create_weight_count": 1,
                                            "build_scorer": 7751
                                        }
                                    },
                                    {
                                        "type": "BoostQuery",
                                        "description": "(title:mountain title:lake)^0.9",
                                        "time_in_nanos": 196085,
                                        "breakdown": {
                                            "set_min_competitive_score_count": 0,
                                            "match_count": 0,
                                            "shallow_advance_count": 4,
                                            "set_min_competitive_score": 0,
                                            "next_doc": 0,
                                            "match": 0,
                                            "next_doc_count": 0,
                                            "score_count": 1,
                                            "compute_max_score_count": 4,
                                            "compute_max_score": 7833,
                                            "advance": 2417,
                                            "advance_count": 2,
                                            "count_weight_count": 0,
                                            "score": 1667,
                                            "build_scorer_count": 5,
                                            "create_weight": 101084,
                                            "shallow_advance": 2543,
                                            "count_weight": 0,
                                            "create_weight_count": 1,
                                            "build_scorer": 80541
                                        },
                                        "children": [
                                            {
                                                "type": "TermQuery",
                                                "description": "title:mountain",
                                                "time_in_nanos": 44667,
                                                "breakdown": {
                                                    "set_min_competitive_score_count": 0,
                                                    "match_count": 0,
                                                    "shallow_advance_count": 0,
                                                    "set_min_competitive_score": 0,
                                                    "next_doc": 0,
                                                    "match": 0,
                                                    "next_doc_count": 0,
                                                    "score_count": 0,
                                                    "compute_max_score_count": 0,
                                                    "compute_max_score": 0,
                                                    "advance": 0,
                                                    "advance_count": 0,
                                                    "count_weight_count": 0,
                                                    "score": 0,
                                                    "build_scorer_count": 3,
                                                    "create_weight": 29125,
                                                    "shallow_advance": 0,
                                                    "count_weight": 0,
                                                    "create_weight_count": 1,
                                                    "build_scorer": 15542
                                                }
                                            },
                                            {
                                                "type": "TermQuery",
                                                "description": "title:lake",
                                                "time_in_nanos": 127373,
                                                "breakdown": {
                                                    "set_min_competitive_score_count": 0,
                                                    "match_count": 0,
                                                    "shallow_advance_count": 4,
                                                    "set_min_competitive_score": 0,
                                                    "next_doc": 0,
                                                    "match": 0,
                                                    "next_doc_count": 0,
                                                    "score_count": 1,
                                                    "compute_max_score_count": 4,
                                                    "compute_max_score": 7500,
                                                    "advance": 2166,
                                                    "advance_count": 2,
                                                    "count_weight_count": 0,
                                                    "score": 1583,
                                                    "build_scorer_count": 5,
                                                    "create_weight": 64792,
                                                    "shallow_advance": 2124,
                                                    "count_weight": 0,
                                                    "create_weight_count": 1,
                                                    "build_scorer": 49208
                                                }
                                            }
                                        ]
                                    }
                                ]
                            }
                        ],
                        "rewrite_time": 80166,
                        "collector": [
                            {
                                "name": "SimpleTopScoreDocCollector",
                                "reason": "search_top_hits",
                                "time_in_nanos": 15042
                            }
                        ]
                    }
                ],
                "aggregations": [],
                "fetch": {
                    "type": "fetch",
                    "description": "",
                    "time_in_nanos": 227541,
                    "breakdown": {
                        "load_stored_fields": 89584,
                        "load_source": 13333,
                        "load_stored_fields_count": 4,
                        "next_reader_count": 3,
                        "load_source_count": 4,
                        "next_reader": 20792
                    },
                    "debug": {
                        "stored_fields": [
                            "_id",
                            "_routing",
                            "_source"
                        ]
                    },
                    "children": [
                        {
                            "type": "FetchSourcePhase",
                            "description": "",
                            "time_in_nanos": 9209,
                            "breakdown": {
                                "process_count": 4,
                                "process": 8958,
                                "next_reader": 251,
                                "next_reader_count": 3
                            },
                            "debug": {
                                "fast_path": 4
                            }
                        },
                        {
                            "type": "StoredFieldsPhase",
                            "description": "",
                            "time_in_nanos": 5582,
                            "breakdown": {
                                "process_count": 4,
                                "process": 5165,
                                "next_reader": 417,
                                "next_reader_count": 3
                            }
                        }
                    ]
                }
            }
        ]
    }
}
```

@javanna && @jdconrad let me know if this seems logical to y'all :)

Reply:
@benwtrent This seems like the best solution to me for the most part for profiling. The only thing I'm concerned about is the rewrite time is combined (if I'm looking at this output correctly) for all knn searches. I wonder if that should be split per knn query as I think it would be useful to know which particular knn query is taking the most time.

Reply:
We could serialize a `List<QueryProfileShardResult>`. I would have to add some BWC serialization logic to make sure we don't break things.

Reply:
That makes sense to me if it's not a huge amount of work to add. Otherwise, as a work around a user can profile the individual knn queries as separate queries to get the breakdown.

Reply:
Yep, here what it looks like now for multiple KNN queries:
```
"profile": {
        "shards": [
            {
                "id": "[ejTy6TXwRiqx3a0_k4SoGA][image-index][0]",
                "dfs": {
                    "statistics": {
                        "type": "statistics",
                        "description": "collect term statistics",
                        "time_in_nanos": 292709,
                        "breakdown": {
                            "term_statistics": 2500,
                            "collection_statistics": 22958,
                            "collection_statistics_count": 2,
                            "create_weight": 287583,
                            "term_statistics_count": 1,
                            "rewrite_count": 0,
                            "create_weight_count": 1,
                            "rewrite": 0
                        }
                    },
                    "knn": [
                        {
                            "query": [
                                {
                                    "type": "BoostQuery",
                                    "description": "(DocAndScore[50])^0.1",
                                    "time_in_nanos": 18250,
                                    "breakdown": {
                                        "set_min_competitive_score_count": 0,
                                        "match_count": 0,
                                        "shallow_advance_count": 0,
                                        "set_min_competitive_score": 0,
                                        "next_doc": 1042,
                                        "match": 0,
                                        "next_doc_count": 3,
                                        "score_count": 3,
                                        "compute_max_score_count": 0,
                                        "compute_max_score": 0,
                                        "advance": 3375,
                                        "advance_count": 4,
                                        "count_weight_count": 0,
                                        "score": 958,
                                        "build_scorer_count": 8,
                                        "create_weight": 2833,
                                        "shallow_advance": 0,
                                        "count_weight": 0,
                                        "create_weight_count": 1,
                                        "build_scorer": 10042
                                    }
                                }
                            ],
                            "rewrite_time": 2547875,
                            "collector": [
                                {
                                    "name": "SimpleTopScoreDocCollector",
                                    "reason": "search_top_hits",
                                    "time_in_nanos": 16791
                                }
                            ]
                        },
                        {
                            "query": [
                                {
                                    "type": "BoostQuery",
                                    "description": "(DocAndScore[10])^0.5",
                                    "time_in_nanos": 16292,
                                    "breakdown": {
                                        "set_min_competitive_score_count": 0,
                                        "match_count": 0,
                                        "shallow_advance_count": 0,
                                        "set_min_competitive_score": 0,
                                        "next_doc": 875,
                                        "match": 0,
                                        "next_doc_count": 3,
                                        "score_count": 3,
                                        "compute_max_score_count": 0,
                                        "compute_max_score": 0,
                                        "advance": 3625,
                                        "advance_count": 4,
                                        "count_weight_count": 0,
                                        "score": 875,
                                        "build_scorer_count": 8,
                                        "create_weight": 1625,
                                        "shallow_advance": 0,
                                        "count_weight": 0,
                                        "create_weight_count": 1,
                                        "build_scorer": 9292
                                    }
                                }
                            ],
                            "rewrite_time": 427500,
                            "collector": [
                                {
                                    "name": "SimpleTopScoreDocCollector",
                                    "reason": "search_top_hits",
                                    "time_in_nanos": 13751
                                }
                            ]
                        }
                    ]
                },
                "searches": [
                    {
                        "query": [
                            {
                                "type": "BooleanQuery",
                                "description": "ScoreAndDocQuery (title:mountain title:lake)^0.9",
                                "time_in_nanos": 673459,
                                "breakdown": {
                                    "set_min_competitive_score_count": 0,
                                    "match_count": 1,
                                    "shallow_advance_count": 0,
                                    "set_min_competitive_score": 0,
                                    "next_doc": 20375,
                                    "match": 375,
                                    "next_doc_count": 4,
                                    "score_count": 4,
                                    "compute_max_score_count": 0,
                                    "compute_max_score": 0,
                                    "advance": 40833,
                                    "advance_count": 4,
                                    "count_weight_count": 0,
                                    "score": 7583,
                                    "build_scorer_count": 8,
                                    "create_weight": 226417,
                                    "shallow_advance": 0,
                                    "count_weight": 0,
                                    "create_weight_count": 1,
                                    "build_scorer": 377876
                                },
                                "children": [
                                    {
                                        "type": "KnnScoreDocQuery",
                                        "description": "ScoreAndDocQuery",
                                        "time_in_nanos": 37499,
                                        "breakdown": {
                                            "set_min_competitive_score_count": 0,
                                            "match_count": 0,
                                            "shallow_advance_count": 4,
                                            "set_min_competitive_score": 0,
                                            "next_doc": 1375,
                                            "match": 0,
                                            "next_doc_count": 3,
                                            "score_count": 4,
                                            "compute_max_score_count": 4,
                                            "compute_max_score": 4001,
                                            "advance": 6499,
                                            "advance_count": 5,
                                            "count_weight_count": 0,
                                            "score": 1249,
                                            "build_scorer_count": 12,
                                            "create_weight": 1833,
                                            "shallow_advance": 3874,
                                            "count_weight": 0,
                                            "create_weight_count": 1,
                                            "build_scorer": 18668
                                        }
                                    },
                                    {
                                        "type": "BoostQuery",
                                        "description": "(title:mountain title:lake)^0.9",
                                        "time_in_nanos": 406791,
                                        "breakdown": {
                                            "set_min_competitive_score_count": 0,
                                            "match_count": 0,
                                            "shallow_advance_count": 4,
                                            "set_min_competitive_score": 0,
                                            "next_doc": 0,
                                            "match": 0,
                                            "next_doc_count": 0,
                                            "score_count": 1,
                                            "compute_max_score_count": 4,
                                            "compute_max_score": 14875,
                                            "advance": 3625,
                                            "advance_count": 2,
                                            "count_weight_count": 0,
                                            "score": 3208,
                                            "build_scorer_count": 6,
                                            "create_weight": 208584,
                                            "shallow_advance": 4959,
                                            "count_weight": 0,
                                            "create_weight_count": 1,
                                            "build_scorer": 171540
                                        },
                                        "children": [
                                            {
                                                "type": "TermQuery",
                                                "description": "title:mountain",
                                                "time_in_nanos": 89416,
                                                "breakdown": {
                                                    "set_min_competitive_score_count": 0,
                                                    "match_count": 0,
                                                    "shallow_advance_count": 0,
                                                    "set_min_competitive_score": 0,
                                                    "next_doc": 0,
                                                    "match": 0,
                                                    "next_doc_count": 0,
                                                    "score_count": 0,
                                                    "compute_max_score_count": 0,
                                                    "compute_max_score": 0,
                                                    "advance": 0,
                                                    "advance_count": 0,
                                                    "count_weight_count": 0,
                                                    "score": 0,
                                                    "build_scorer_count": 4,
                                                    "create_weight": 56333,
                                                    "shallow_advance": 0,
                                                    "count_weight": 0,
                                                    "create_weight_count": 1,
                                                    "build_scorer": 33083
                                                }
                                            },
                                            {
                                                "type": "TermQuery",
                                                "description": "title:lake",
                                                "time_in_nanos": 265374,
                                                "breakdown": {
                                                    "set_min_competitive_score_count": 0,
                                                    "match_count": 0,
                                                    "shallow_advance_count": 4,
                                                    "set_min_competitive_score": 0,
                                                    "next_doc": 0,
                                                    "match": 0,
                                                    "next_doc_count": 0,
                                                    "score_count": 1,
                                                    "compute_max_score_count": 4,
                                                    "compute_max_score": 13792,
                                                    "advance": 3042,
                                                    "advance_count": 2,
                                                    "count_weight_count": 0,
                                                    "score": 2958,
                                                    "build_scorer_count": 6,
                                                    "create_weight": 139416,
                                                    "shallow_advance": 3958,
                                                    "count_weight": 0,
                                                    "create_weight_count": 1,
                                                    "build_scorer": 102208
                                                }
                                            }
                                        ]
                                    }
                                ]
                            }
                        ],
                        "rewrite_time": 180875,
                        "collector": [
                            {
                                "name": "SimpleTopScoreDocCollector",
                                "reason": "search_top_hits",
                                "time_in_nanos": 29499
                            }
                        ]
                    }
                ],
                "aggregations": [],
                "fetch": {
                    "type": "fetch",
                    "description": "",
                    "time_in_nanos": 701041,
                    "breakdown": {
                        "load_stored_fields": 308208,
                        "load_source": 28582,
                        "load_stored_fields_count": 4,
                        "next_reader_count": 4,
                        "load_source_count": 4,
                        "next_reader": 71666
                    },
                    "debug": {
                        "stored_fields": [
                            "_id",
                            "_routing",
                            "_source"
                        ]
                    },
                    "children": [
                        {
                            "type": "FetchSourcePhase",
                            "description": "",
                            "time_in_nanos": 35332,
                            "breakdown": {
                                "process_count": 4,
                                "process": 34666,
                                "next_reader": 666,
                                "next_reader_count": 4
                            },
                            "debug": {
                                "fast_path": 4
                            }
                        },
                        {
                            "type": "StoredFieldsPhase",
                            "description": "",
                            "time_in_nanos": 21541,
                            "breakdown": {
                                "process_count": 4,
                                "process": 20292,
                                "next_reader": 1249,
                                "next_reader_count": 4
                            }
                        }
                    ]
                }
            }
        ]
    }
```

Here is just a single KNN query (note, it is still an array, I think this is good).
```
"profile": {
        "shards": [
            {
                "id": "[ejTy6TXwRiqx3a0_k4SoGA][image-index][0]",
                "dfs": {
                    "statistics": {
                        "type": "statistics",
                        "description": "collect term statistics",
                        "time_in_nanos": 1312750,
                        "breakdown": {
                            "term_statistics": 4833,
                            "collection_statistics": 42000,
                            "collection_statistics_count": 2,
                            "create_weight": 1302833,
                            "term_statistics_count": 1,
                            "rewrite_count": 0,
                            "create_weight_count": 1,
                            "rewrite": 0
                        }
                    },
                    "knn": [
                        {
                            "query": [
                                {
                                    "type": "BoostQuery",
                                    "description": "(DocAndScore[50])^0.1",
                                    "time_in_nanos": 46375,
                                    "breakdown": {
                                        "set_min_competitive_score_count": 0,
                                        "match_count": 0,
                                        "shallow_advance_count": 0,
                                        "set_min_competitive_score": 0,
                                        "next_doc": 1332,
                                        "match": 0,
                                        "next_doc_count": 4,
                                        "score_count": 4,
                                        "compute_max_score_count": 0,
                                        "compute_max_score": 0,
                                        "advance": 8583,
                                        "advance_count": 4,
                                        "count_weight_count": 0,
                                        "score": 1460,
                                        "build_scorer_count": 8,
                                        "create_weight": 4375,
                                        "shallow_advance": 0,
                                        "count_weight": 0,
                                        "create_weight_count": 1,
                                        "build_scorer": 30625
                                    }
                                }
                            ],
                            "rewrite_time": 568916,
                            "collector": [
                                {
                                    "name": "SimpleTopScoreDocCollector",
                                    "reason": "search_top_hits",
                                    "time_in_nanos": 68917
                                }
                            ]
                        }
                    ]
                },
                "searches": [
                    {
                        "query": [
                            {
                                "type": "BooleanQuery",
                                "description": "ScoreAndDocQuery (title:mountain title:lake)^0.9",
                                "time_in_nanos": 950872,
                                "breakdown": {
                                    "set_min_competitive_score_count": 0,
                                    "match_count": 1,
                                    "shallow_advance_count": 0,
                                    "set_min_competitive_score": 0,
                                    "next_doc": 16541,
                                    "match": 458,
                                    "next_doc_count": 4,
                                    "score_count": 4,
                                    "compute_max_score_count": 0,
                                    "compute_max_score": 0,
                                    "advance": 42458,
                                    "advance_count": 4,
                                    "count_weight_count": 0,
                                    "score": 7624,
                                    "build_scorer_count": 8,
                                    "create_weight": 497250,
                                    "shallow_advance": 0,
                                    "count_weight": 0,
                                    "create_weight_count": 1,
                                    "build_scorer": 386541
                                },
                                "children": [
                                    {
                                        "type": "KnnScoreDocQuery",
                                        "description": "ScoreAndDocQuery",
                                        "time_in_nanos": 42541,
                                        "breakdown": {
                                            "set_min_competitive_score_count": 0,
                                            "match_count": 0,
                                            "shallow_advance_count": 4,
                                            "set_min_competitive_score": 0,
                                            "next_doc": 1125,
                                            "match": 0,
                                            "next_doc_count": 3,
                                            "score_count": 4,
                                            "compute_max_score_count": 4,
                                            "compute_max_score": 2417,
                                            "advance": 5665,
                                            "advance_count": 5,
                                            "count_weight_count": 0,
                                            "score": 1248,
                                            "build_scorer_count": 12,
                                            "create_weight": 2834,
                                            "shallow_advance": 4127,
                                            "count_weight": 0,
                                            "create_weight_count": 1,
                                            "build_scorer": 25125
                                        }
                                    },
                                    {
                                        "type": "BoostQuery",
                                        "description": "(title:mountain title:lake)^0.9",
                                        "time_in_nanos": 628874,
                                        "breakdown": {
                                            "set_min_competitive_score_count": 0,
                                            "match_count": 0,
                                            "shallow_advance_count": 4,
                                            "set_min_competitive_score": 0,
                                            "next_doc": 0,
                                            "match": 0,
                                            "next_doc_count": 0,
                                            "score_count": 1,
                                            "compute_max_score_count": 4,
                                            "compute_max_score": 14375,
                                            "advance": 3667,
                                            "advance_count": 2,
                                            "count_weight_count": 0,
                                            "score": 3458,
                                            "build_scorer_count": 6,
                                            "create_weight": 431750,
                                            "shallow_advance": 4833,
                                            "count_weight": 0,
                                            "create_weight_count": 1,
                                            "build_scorer": 170791
                                        },
                                        "children": [
                                            {
                                                "type": "TermQuery",
                                                "description": "title:mountain",
                                                "time_in_nanos": 139125,
                                                "breakdown": {
                                                    "set_min_competitive_score_count": 0,
                                                    "match_count": 0,
                                                    "shallow_advance_count": 0,
                                                    "set_min_competitive_score": 0,
                                                    "next_doc": 0,
                                                    "match": 0,
                                                    "next_doc_count": 0,
                                                    "score_count": 0,
                                                    "compute_max_score_count": 0,
                                                    "compute_max_score": 0,
                                                    "advance": 0,
                                                    "advance_count": 0,
                                                    "count_weight_count": 0,
                                                    "score": 0,
                                                    "build_scorer_count": 4,
                                                    "create_weight": 111208,
                                                    "shallow_advance": 0,
                                                    "count_weight": 0,
                                                    "create_weight_count": 1,
                                                    "build_scorer": 27917
                                                }
                                            },
                                            {
                                                "type": "TermQuery",
                                                "description": "title:lake",
                                                "time_in_nanos": 373790,
                                                "breakdown": {
                                                    "set_min_competitive_score_count": 0,
                                                    "match_count": 0,
                                                    "shallow_advance_count": 4,
                                                    "set_min_competitive_score": 0,
                                                    "next_doc": 0,
                                                    "match": 0,
                                                    "next_doc_count": 0,
                                                    "score_count": 1,
                                                    "compute_max_score_count": 4,
                                                    "compute_max_score": 13667,
                                                    "advance": 3333,
                                                    "advance_count": 2,
                                                    "count_weight_count": 0,
                                                    "score": 3250,
                                                    "build_scorer_count": 6,
                                                    "create_weight": 263708,
                                                    "shallow_advance": 4167,
                                                    "count_weight": 0,
                                                    "create_weight_count": 1,
                                                    "build_scorer": 85665
                                                }
                                            }
                                        ]
                                    }
                                ]
                            }
                        ],
                        "rewrite_time": 167124,
                        "collector": [
                            {
                                "name": "SimpleTopScoreDocCollector",
                                "reason": "search_top_hits",
                                "time_in_nanos": 31082
                            }
                        ]
                    }
                ],
                "aggregations": [],
                "fetch": {
                    "type": "fetch",
                    "description": "",
                    "time_in_nanos": 433708,
                    "breakdown": {
                        "load_stored_fields": 95208,
                        "load_source": 30166,
                        "load_stored_fields_count": 4,
                        "next_reader_count": 4,
                        "load_source_count": 4,
                        "next_reader": 47749
                    },
                    "debug": {
                        "stored_fields": [
                            "_id",
                            "_routing",
                            "_source"
                        ]
                    },
                    "children": [
                        {
                            "type": "FetchSourcePhase",
                            "description": "",
                            "time_in_nanos": 19499,
                            "breakdown": {
                                "process_count": 4,
                                "process": 18916,
                                "next_reader": 583,
                                "next_reader_count": 4
                            },
                            "debug": {
                                "fast_path": 4
                            }
                        },
                        {
                            "type": "StoredFieldsPhase",
                            "description": "",
                            "time_in_nanos": 13625,
                            "breakdown": {
                                "process_count": 4,
                                "process": 12833,
                                "next_reader": 792,
                                "next_reader_count": 4
                            }
                        }
                    ]
                }
            }
        ]
    }
```

For BWC, I will combine the results like I was doing previously here when serializing to an older coordinator. This way we at least report the total numbers.

Reply:
This lgtm!

Reply:
OK, pushed a commit allowing for multiple query profile shard results.

Reply:
LGTM too.